File: src/lib/prompt-assembler.ts
/**
 * Context engineering for asset consistency (server-only).
 *
 * Turns a raw user prompt + the assets/uploads it references into a structured,
 * role-labeled payload: a text instruction that lists every locked asset and
 * keeps the SCENE literal, plus per-asset GROUPS of reference images. Grouping
 * several images under one tag (e.g. 4 angles of a face) and binding each group
 * to its @tag is what actually holds identity across generations.
 */

import { readImageAsBase64 } from "./save-media";
import { detectReferenceRole, identityCrops, prepReference } from "./middleware/image-prep";
import { parseAssetSlugs, parseMentionIndices } from "./mentions";
import {
  buildReferenceLegend,
  buildShotInstruction,
  parseRefRoles,
  roleHeader,
  type LegendEntry,
  type RefRole,
} from "./shot-spec";
import type { Asset, AssetKind } from "./types";

export interface AssembledImage {
  mimeType: string;
  data: string; // base64 (no data: prefix)
}

export interface AssembledGroup {
  tag: string; // "@priya" | "@img1"
  /** Header shown right before this group's images, binding them to the tag. */
  header: string;
  images: AssembledImage[];
  /** True when this group is identity ground truth for a person (character
   *  asset, SUBJECT uploads, or a person-classified @imgN upload). */
  identity?: boolean;
  /** Identity tiles: face/sheet-panel crops of this group's images, sent as
   *  EXTRA images (Gemini ingests every image as one flat ~258-token tile, so
   *  a wide character sheet carries almost no facial detail on its own).
   *  Kept separate from `images` so providers can budget them — user images
   *  are never dropped to make room for tiles. */
  tiles?: AssembledImage[];
}

export interface AssembledPrompt {
  /** Full structured text for the first user content part. Always the raw
   *  prompt — never overwritten, in either mode. */
  instruction: string;
  /** Structured text (role legend + literal SCENE + framing/negative codas),
   *  set only when PROMPT_SHOT_SPEC=1. Already contains its own "SCENE:"
   *  prefix — providers must use this verbatim instead of re-wrapping
   *  `instruction` when present. */
  shotInstruction?: string;
  /** Reference-image groups, in the order they should be sent. */
  groups: AssembledGroup[];
  /** Best available face crop for judging generated frames (best-of-N). */
  judgeFace?: AssembledImage;
}

const KIND_LABEL: Record<AssetKind, string> = {
  character: "CHARACTER",
  outfit: "OUTFIT",
  location: "LOCATION",
  style: "STYLE",
  prop: "PROP",
};

/** Maps a named-asset kind onto the shot-spec RefRole vocabulary. */
const ASSET_KIND_TO_ROLE: Record<AssetKind, RefRole> = {
  character: "person",
  outfit: "outfit",
  location: "location",
  style: "style",
  prop: "prop",
};

const KIND_RULE: Record<AssetKind, string> = {
  character:
    "reproduce this exact person with photographic fidelity — identical face shape and bone structure, jawline, cheekbones, hairline, eye shape/size/spacing and color, eyebrows, nose, lips, ears, skin tone and texture (keep moles, scars, freckles, wrinkles), facial hair, hairstyle, body build and apparent age; unmistakably the SAME individual, never a lookalike, and never beautified, slimmed, de-aged or idealized",
  outfit:
    "reproduce this exact outfit — same garments, cut, fit, fabric, colors, patterns, trims and details, plus any jewelry/accessories shown with it",
  location:
    "reproduce this exact place — same architecture, layout, materials, signage, furnishing and mood",
  style:
    "match this exact visual style — same rendering, palette, grain and lighting treatment",
  prop: "reproduce this exact object — same shape, colors, materials, markings and wear",
};

async function readAll(refs: string[]): Promise<AssembledImage[]> {
  const out: AssembledImage[] = [];
  for (const ref of refs) {
    try {
      const { mimeType, data } = await readImageAsBase64(ref);
      // Middleware: cap oversized refs (Higgsfield-style resize preprocessing).
      out.push(await prepReference(mimeType, data));
    } catch {
      /* skip unreadable image */
    }
  }
  return out;
}

/**
 * Identity tiling: crop the face (and character-sheet panels) into their OWN
 * images, multiplying an identity ref's visual bandwidth 3–4×. Detection
 * gates non-person refs (locations, sets) so no wrong face is injected.
 * Fail-open: [] when nothing is found (or FACE_CROP_MIDDLEWARE=0).
 */
async function faceCrops(
  images: AssembledImage[],
  tag: string,
  limit = 2
): Promise<AssembledImage[]> {
  const out: AssembledImage[] = [];
  for (const img of images.slice(0, limit)) {
    if (out.length >= 3) break;
    out.push(...(await identityCrops(img.mimeType, img.data, 3 - out.length)));
  }
  console.log(
    `[middleware] ${tag}: identity tile${out.length === 1 ? "" : "s"} ${
      out.length ? `added (${out.length})` : "none (not a person ref / no face)"
    }`
  );
  return out;
}

/**
 * Resolve the RefRole for an @imgN / SUBJECT upload when PROMPT_SHOT_SPEC=1.
 * Precedence: prompt-text keyword scan (parseRefRoles) wins when present — it
 * is the user's explicit binding contract; PROMPT_ROLE_DETECT=1 is consulted
 * as a fallback AND a cross-check (a mismatch is logged, never auto-"fixed",
 * so ref/tag ordering mistakes surface to humans instead of being silently
 * reconciled); otherwise fall back to today's identity signal (tiles found →
 * person, else object). Fail-open: detection errors resolve to null upstream.
 */
async function resolveUploadRole(
  tag: string,
  firstImage: AssembledImage,
  tiles: AssembledImage[],
  textRoles: Map<string, RefRole>,
  roleDetectOn: boolean
): Promise<RefRole> {
  const textRole = textRoles.get(tag);
  if (textRole) {
    if (roleDetectOn) {
      // Cross-check is diagnostic only — don't spend hot-path latency on it
      // (detection is ~3s/ref, sequential); let it resolve during generation.
      void detectReferenceRole(firstImage.mimeType, firstImage.data)
        .then((detected) => {
          if (detected && detected !== textRole) {
            console.log(
              `[shot-spec] WARN role mismatch for ${tag}: prompt text says ` +
                `"${textRole}", detection says "${detected}" — using the prompt-text role.`
            );
          }
        })
        .catch(() => {});
    }
    return textRole;
  }
  const detected = roleDetectOn
    ? await detectReferenceRole(firstImage.mimeType, firstImage.data)
    : null;
  return detected ?? (tiles.length > 0 ? "person" : "object");
}

/**
 * Build the assembled payload.
 * @param prompt  raw user prompt (kept literal in the SCENE block)
 * @param assets  all known assets (referenced ones are matched by @slug)
 * @param uploads ad-hoc data-URL uploads, 1-based for @imgN
 * @param opts    aspectRatio, used only to gate the wide-AR framing coda when
 *                PROMPT_SHOT_SPEC=1; medium ("image" | "video", default
 *                "image") selects the shot-spec's framing/AVOID wording for
 *                the Omni video path. Optional/omitted ⇒ today's behavior.
 */
export async function assemblePrompt(
  prompt: string,
  assets: Asset[],
  uploads: string[],
  opts?: { aspectRatio?: string; medium?: "image" | "video" }
): Promise<AssembledPrompt> {
  const groups: AssembledGroup[] = [];
  const assetLines: string[] = [];

  // Shot-spec mode (PROMPT_SHOT_SPEC=1): role-aware headers + a reference
  // legend + a structured shotInstruction. Off = untouched, byte-identical.
  const shotSpecOn = process.env.PROMPT_SHOT_SPEC === "1";
  const roleDetectOn = process.env.PROMPT_ROLE_DETECT === "1";
  const textRoles = shotSpecOn ? parseRefRoles(prompt) : new Map<string, RefRole>();
  const legendEntries: LegendEntry[] = [];

  // 1) Named asset references (@slug) — only those actually mentioned.
  const slugs = parseAssetSlugs(prompt);
  const bySlug = new Map(assets.map((a) => [a.slug, a]));
  for (const slug of slugs) {
    const asset = bySlug.get(slug);
    if (!asset || !asset.images.length) continue;
    const images = await readAll(asset.images);
    if (!images.length) continue;
    const label = KIND_LABEL[asset.kind];
    const desc = asset.description ? ` — ${asset.description}` : "";
    assetLines.push(
      `- @${asset.slug} → ${label} "${asset.name}"${desc}. Rule: ${KIND_RULE[asset.kind]}.`
    );
    const isCharacter = asset.kind === "character";
    const tag = `@${asset.slug}`;
    let header =
      `@${asset.slug} — ${label} "${asset.name}" ` +
      `(${images.length} reference image${images.length > 1 ? "s" : ""}; ` +
      `${KIND_RULE[asset.kind]}):`;
    if (shotSpecOn) {
      const role = ASSET_KIND_TO_ROLE[asset.kind];
      header = roleHeader(tag, role, images.length);
      legendEntries.push({ tag, role, isPerson: role === "person" });
    }
    groups.push({
      tag,
      header,
      images,
      identity: isCharacter,
      tiles: isCharacter ? await faceCrops(images, tag) : undefined,
    });
  }

  // 2) Ad-hoc uploads.
  //  - If the prompt tags images distinctly (@img1, @img2 …) the user is
  //    pointing at DIFFERENT subjects → keep each separate.
  //  - Otherwise (the common case: a few photos of ONE person, no tags) treat
  //    ALL uploads as multiple angles of the SAME person, so identity locks
  //    instead of the model reading them as different people.
  let hasIdentityRef = false;
  if (uploads.length) {
    const tagged = parseMentionIndices(prompt).filter((n) => n <= uploads.length);
    if (tagged.length > 1) {
      for (const n of tagged) {
        const images = await readAll([uploads[n - 1]]);
        if (!images.length) continue;
        hasIdentityRef = true;
        const tag = `@img${n}`;
        assetLines.push(
          `- @img${n} → REFERENCE. Rule: reproduce the tagged subject from this ` +
            `image exactly — if it is a person, the identical face (bone ` +
            `structure, eyes, nose, lips, skin tone/texture, marks, hairline, ` +
            `facial hair, apparent age), hairstyle, build and visible ` +
            `outfit/jewelry, changed only where the SCENE explicitly says so. ` +
            `Never blend @img${n} with any other reference.`
        );
        // Person detection decides identity: a face/sheet upload gets tiles;
        // outfit/location/style uploads yield none and stay non-identity.
        const tiles = await faceCrops(images, tag, 1);
        let header =
          `${tag} — REFERENCE (reproduce this subject exactly; if a person, ` +
          `the same individual — identical facial features, never a lookalike):`;
        if (shotSpecOn) {
          const role = await resolveUploadRole(tag, images[0], tiles, textRoles, roleDetectOn);
          header = roleHeader(tag, role, images.length);
          legendEntries.push({ tag, role, isPerson: role === "person" });
        }
        groups.push({
          tag,
          header,
          images,
          identity: tiles.length > 0,
          tiles: tiles.length ? tiles : undefined,
        });
      }
    } else {
      const images = await readAll(uploads);
      if (images.length) {
        hasIdentityRef = true;
        const many = images.length > 1;
        assetLines.push(
          `- SUBJECT → the person in the reference photo${many ? "s" : ""} below. ` +
            `Rule: the main subject in the output MUST be this exact same person — ` +
            `identical face, hairstyle, build and visible outfit/jewelry unless ` +
            `the SCENE explicitly changes them.`
        );
        const tiles = await faceCrops(images, "SUBJECT");
        let header = many
          ? `SUBJECT — ${images.length} reference photos of the SAME person ` +
            `(different angles/lighting). Reconstruct ONE consistent identity ` +
            `from all of them; the generated person's face MUST match exactly — ` +
            `same bone structure, jawline, hairline, eye shape/spacing and color, ` +
            `eyebrows, nose, lips, skin tone/texture (keep moles, scars, ` +
            `freckles), facial hair and apparent age. Keep their hairstyle, ` +
            `build and worn outfit/jewelry unless the SCENE explicitly changes ` +
            `them. A recognizable match, never a lookalike — never beautified ` +
            `or idealized:`
          : `SUBJECT — reference photo of the person. The generated person's face ` +
            `MUST be this exact same individual — same bone structure, jawline, ` +
            `hairline, eye shape/spacing and color, eyebrows, nose, lips, skin ` +
            `tone/texture (keep moles, scars, freckles), facial hair and ` +
            `apparent age. Keep their hairstyle, build and worn outfit/jewelry ` +
            `unless the SCENE explicitly changes them. A recognizable match, ` +
            `not a lookalike — never beautified or idealized:`;
        if (shotSpecOn) {
          // SUBJECT is always the untagged-upload identity ref → role person.
          header = roleHeader("SUBJECT", "person", images.length);
          legendEntries.push({ tag: "SUBJECT", role: "person", isPerson: true });
        }
        groups.push({
          tag: "SUBJECT",
          tiles: tiles.length ? tiles : undefined,
          header,
          images,
          identity: true,
        });
      }
    }
  }

  // 3) The instruction stays the RAW prompt. Wrapping it in big instruction
  // blocks gets rendered literally (the model drew a movie camera on set);
  // reference binding lives in the group headers instead.
  // judgeFace: identityCrops returns the face close-up first when one exists —
  // the best ground truth for scoring generated frames (best-of-N).
  const judgeFace = groups.find((g) => g.identity && g.tiles?.length)?.tiles?.[0];

  // shotInstruction: the structured shape (legend + literal SCENE + framing/
  // negative codas) — built only in shot-spec mode, never replacing `prompt`.
  let shotInstruction: string | undefined;
  if (shotSpecOn) {
    const legend = buildReferenceLegend(legendEntries);
    shotInstruction = buildShotInstruction({
      rawPrompt: prompt,
      legend,
      aspectRatio: opts?.aspectRatio || "1:1",
      medium: opts?.medium,
    });
  }

  return { instruction: prompt, shotInstruction, groups, judgeFace };
}

File: backend/apps/generation/prompt_assembler.py
"""Port of src/lib/prompt-assembler.js — context engineering for asset
consistency (server-only). See that file's header for the full design
rationale: role-labeled reference GROUPS bound to @tags via headers, plus
identity TILES (face crops sent as extra images) so Gemini — which ingests
every image as one flat ~258-token tile — gets real facial bandwidth.

This is the single most measured file in the migration. Every branch here
mirrors the TS file's branch for branch; if you change one side, change
the other in the same change.
"""

import os

from apps.media import storage

from . import shot_spec as ss
from .image_prep import analyze_identity_reference, detect_reference_role, prep_reference
from .mentions import parse_asset_slugs, parse_mention_indices

KIND_LABEL = {
    "character": "CHARACTER",
    "outfit": "OUTFIT",
    "location": "LOCATION",
    "style": "STYLE",
    "prop": "PROP",
}

ASSET_KIND_TO_ROLE = {
    "character": "person",
    "outfit": "outfit",
    "location": "location",
    "style": "style",
    "prop": "prop",
}

KIND_RULE = {
    "character": (
        "reproduce this exact person with exact fidelity to the reference, in the SAME medium and rendering "
        "style as the reference (photographic, illustrated, anime, cel-shaded, 3D, painterly or otherwise — "
        "never convert it to a different medium, and never add realism the reference does not have) — identical "
        "face shape and bone structure, jawline, cheekbones, hairline, eye shape/size/spacing and color, "
        "eyebrows, nose, lips, ears, facial hair, hairstyle, body build and apparent age, plus the same "
        "distinguishing marks the reference shows; where the reference is photographic, also keep real skin "
        "tone and texture (moles, scars, freckles, wrinkles); unmistakably the SAME individual, never a "
        "lookalike, and never beautified, slimmed, de-aged or idealized relative to the reference"
    ),
    "outfit": (
        "reproduce this exact outfit — same garments, cut, fit, fabric, colors, patterns, trims and details, "
        "plus any jewelry/accessories shown with it"
    ),
    "location": "reproduce this exact place — same architecture, layout, materials, signage, furnishing and mood",
    "style": "match this exact visual style — same rendering, palette, grain and lighting treatment",
    "prop": "reproduce this exact object — same shape, colors, materials, markings and wear",
}


def _read_all(refs: list[str]) -> list[dict]:
    out = []
    for ref in refs:
        try:
            mime_type, data = storage.read_as_base64(ref)
            out.append(prep_reference(mime_type, data))
        except Exception:
            pass
    return out


def _face_crops(images: list[dict], tag: str, limit: int = 2) -> list[dict]:
    return _analyze_face_crops(images, tag, limit)["tiles"]


def _analyze_face_crops(images: list[dict], tag: str, limit: int = 2) -> dict:
    out: list[dict] = []
    saw_person = saw_non_person = saw_unknown = False
    for img in images[:limit]:
        if len(out) >= 3:
            break
        analysis = analyze_identity_reference(img["mimeType"], img["data"], 3 - len(out))
        out.extend(analysis["crops"])
        if analysis["personReference"] is True:
            saw_person = True
        elif analysis["personReference"] is False:
            saw_non_person = True
        else:
            saw_unknown = True
    print(f"[middleware] {tag}: identity tile{'s' if len(out) != 1 else ''} "
          f"{f'added ({len(out)})' if out else 'none (not a person ref / no face)'}")
    person_reference = True if saw_person else (None if saw_unknown else (False if saw_non_person else None))
    return {"tiles": out, "personReference": person_reference}


def _resolve_upload_role(
    tag: str, first_image: dict, tiles: list[dict], text_roles: dict, role_detect_on: bool
) -> str:
    text_role = text_roles.get(tag)
    if text_role:
        # Cross-check is diagnostic only in the TS version (fire-and-forget,
        # logged on mismatch). Not worth the synchronous latency here either,
        # and we don't have a background-task facility to fire it into — the
        # prompt-text role is authoritative regardless, so skipping the
        # cross-check changes no behavior, only drops a WARN log line.
        return text_role
    detected = detect_reference_role(first_image["mimeType"], first_image["data"]) if role_detect_on else None
    return detected or ("person" if tiles else "object")


def assemble_prompt(prompt: str, assets: list[dict], uploads: list[str], aspect_ratio: str | None = None, medium: str = "image") -> dict:
    """assets: [{"slug","kind","name","description","images"}, ...] (already
    dicts, not ORM instances — callers pass the serialized asset shape).

    Returns {"instruction", "shotInstruction", "groups", "judgeFace"}.
    """
    groups: list[dict] = []
    asset_lines: list[str] = []

    shot_spec_on = os.environ.get("PROMPT_SHOT_SPEC") == "1"
    role_detect_on = os.environ.get("PROMPT_ROLE_DETECT") == "1"
    text_roles = ss.parse_ref_roles(prompt) if shot_spec_on else {}
    legend_entries: list[dict] = []

    # 1) Named asset references (@slug).
    slugs = parse_asset_slugs(prompt)
    by_slug = {a["slug"]: a for a in assets}
    for slug in slugs:
        asset = by_slug.get(slug)
        if not asset or not asset.get("images"):
            continue
        images = _read_all(asset["images"])
        if not images:
            continue
        label = KIND_LABEL[asset["kind"]]
        desc = f" — {asset['description']}" if asset.get("description") else ""
        asset_lines.append(f"- @{asset['slug']} → {label} \"{asset['name']}\"{desc}. Rule: {KIND_RULE[asset['kind']]}.")
        is_character = asset["kind"] == "character"
        tag = f"@{asset['slug']}"
        header = (
            f"@{asset['slug']} — {label} \"{asset['name']}\" "
            f"({len(images)} reference image{'s' if len(images) > 1 else ''}; {KIND_RULE[asset['kind']]}):"
        )
        if shot_spec_on:
            role = ASSET_KIND_TO_ROLE[asset["kind"]]
            header = ss.role_header(tag, role, len(images))
            legend_entries.append({"tag": tag, "role": role, "isPerson": role == "person"})
        groups.append({
            "tag": tag, "header": header, "images": images, "identity": is_character,
            "tiles": _face_crops(images, tag) if is_character else None,
        })

    # 2) Ad-hoc uploads.
    if uploads:
        tagged = [n for n in parse_mention_indices(prompt) if n <= len(uploads)]

        def push_legacy_subject(images: list[dict], prepared_tiles: list[dict] | None = None) -> None:
            if not images:
                return
            many = len(images) > 1
            asset_lines.append(
                f"- SUBJECT → the person in the reference image{'s' if many else ''} below. "
                "Rule: the main subject in the output MUST be this exact same person — "
                "identical face, hairstyle, build and visible outfit/jewelry unless "
                "the SCENE explicitly changes them, rendered in the same medium and "
                "style as the reference."
            )
            tiles = prepared_tiles if prepared_tiles is not None else _face_crops(images, "SUBJECT")
            if many:
                header = (
                    f"SUBJECT — {len(images)} reference images of the SAME person "
                    "(different angles/lighting). Reconstruct ONE consistent identity "
                    "from all of them; the generated person's face MUST match exactly — "
                    "same bone structure, jawline, hairline, eye shape/spacing and color, "
                    "eyebrows, nose, lips, facial hair and apparent age, and where the "
                    "references are photographic also their skin tone/texture (keep "
                    "moles, scars, freckles). Keep their hairstyle, "
                    "build and worn outfit/jewelry unless the SCENE explicitly changes "
                    "them, and render in the same medium and style as the references. "
                    "A recognizable match, never a lookalike — never beautified "
                    "or idealized relative to the references:"
                )
            else:
                header = (
                    "SUBJECT — reference image of the person. The generated person's face "
                    "MUST be this exact same individual — same bone structure, jawline, "
                    "hairline, eye shape/spacing and color, eyebrows, nose, lips, facial "
                    "hair and apparent age, and where the reference is photographic also "
                    "their skin tone/texture (keep moles, scars, freckles). "
                    "Keep their hairstyle, build and worn outfit/jewelry "
                    "unless the SCENE explicitly changes them, and render in the same "
                    "medium and style as the reference. A recognizable match, "
                    "not a lookalike — never beautified or idealized relative to the reference:"
                )
            if shot_spec_on:
                header = ss.role_header("SUBJECT", "person", len(images))
                legend_entries.append({"tag": "SUBJECT", "role": "person", "isPerson": True})
            groups.append({
                "tag": "SUBJECT", "tiles": tiles if tiles else None, "header": header,
                "images": images, "identity": True,
            })

        def push_non_person_group(tag: str, role: str, images: list[dict]) -> None:
            asset_lines.append(f"- {tag} → {role.upper()} REFERENCE.")
            header = ss.role_header(tag, role, len(images))
            legend_entries.append({"tag": tag, "role": role, "isPerson": False})
            groups.append({"tag": tag, "header": header, "images": images, "identity": False})

        if len(tagged) > 1:
            for n in tagged:
                images = _read_all([uploads[n - 1]])
                if not images:
                    continue
                tag = f"@img{n}"
                asset_lines.append(
                    f"- @img{n} → REFERENCE. Rule: reproduce the tagged subject from this "
                    "image exactly — if it is a person, the identical face (bone "
                    "structure, eyes, nose, lips, skin tone/texture, marks, hairline, "
                    "facial hair, apparent age), hairstyle, build and visible "
                    f"outfit/jewelry, changed only where the SCENE explicitly says so. "
                    f"Never blend @img{n} with any other reference."
                )
                tiles = _face_crops(images, tag, 1)
                header = (
                    f"{tag} — REFERENCE (reproduce this subject exactly; if a person, "
                    "the same individual — identical facial features, never a lookalike):"
                )
                identity = len(tiles) > 0
                if shot_spec_on:
                    role = "person" if tiles else _resolve_upload_role(tag, images[0], tiles, text_roles, role_detect_on)
                    header = ss.role_header(tag, role, len(images))
                    legend_entries.append({"tag": tag, "role": role, "isPerson": role == "person"})
                    identity = role == "person"
                groups.append({
                    "tag": tag, "header": header, "images": images, "identity": identity,
                    "tiles": tiles if (identity and tiles) else None,
                })
        elif len(tagged) == 1 and shot_spec_on:
            n = tagged[0]
            tag = f"@img{n}"
            images = _read_all([uploads[n - 1]])
            if images:
                text_role = text_roles.get(tag)
                tiles = _face_crops(images, tag, 1)
                explicit_text_role = text_role is not None and ss.has_explicit_ref_role(prompt, tag, text_role)
                detected_role = None
                if text_role != "person" and not tiles and not explicit_text_role:
                    detected_role = detect_reference_role(images[0]["mimeType"], images[0]["data"])
                role = (
                    "person" if (text_role == "person" or tiles)
                    else (text_role if explicit_text_role else (detected_role or "person"))
                )
                if role == "person":
                    if len(uploads) == 1:
                        push_legacy_subject(images, tiles)
                    else:
                        push_legacy_subject(_read_all(uploads))
                else:
                    push_non_person_group(tag, role, images)
        else:
            images = _read_all(uploads)
            if not shot_spec_on or not images:
                push_legacy_subject(images)
            else:
                tiles = _face_crops(images, "SUBJECT")
                if ss.has_visible_people(prompt) or tiles:
                    push_legacy_subject(images, tiles)
                else:
                    detected_role = detect_reference_role(images[0]["mimeType"], images[0]["data"])
                    if detected_role and detected_role != "person":
                        push_non_person_group("REFERENCE", detected_role, images)
                    else:
                        push_legacy_subject(images, tiles)

    judge_face = None
    for g in groups:
        if g.get("identity") and g.get("tiles"):
            judge_face = g["tiles"][0]
            break

    shot_instruction = None
    if shot_spec_on:
        legend = ss.build_reference_legend(legend_entries)
        shot_instruction = ss.build_shot_instruction(
            raw_prompt=prompt,
            legend=legend,
            aspect_ratio=aspect_ratio or "1:1",
            medium=medium,
            has_person_reference=any(e["isPerson"] for e in legend_entries),
        )

    return {"instruction": prompt, "shotInstruction": shot_instruction, "groups": groups, "judgeFace": judge_face}
